new project created with gradle, kubernetes, jib fails  for task quarkusBuild

```
mvn io.quarkus:quarkus-maven-plugin:1.3.1.Final:create \
    -DprojectGroupId=micro.apps \
    -DprojectArtifactId=greeting-quarkus \
    -DprojectVersion=0.1.0 \
    -DclassName="micro.apps.greeting.GreetingResource" \
    -Dpath="/greeting" \
    -Dextensions="kotlin, resteasy-jsonb, kubernetes, jib" \
    -DbuildTool=gradle
```

```
cd greeting-quarkus
gradle quarkusBuild
```

```
Task :quarkusBuild FAILED
building quarkus runner

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':quarkusBuild'.
> io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.kubernetes.deployment.KubernetesProcessor#build threw an exception: java.lang.UnsupportedOperationException
        at jdk.zipfs/jdk.nio.zipfs.ZipPath.toFile(ZipPath.java:674)
        at io.quarkus.kubernetes.deployment.KubernetesProcessor.createProject(KubernetesProcessor.java:539)
        at io.quarkus.kubernetes.deployment.KubernetesProcessor.build(KubernetesProcessor.java:166)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:938)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:273)
        at io.quarkus.builder.BuildContext$$Lambda$163/0000000000000000.run(Unknown Source)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)

```

I tested with local quarkus build from master, with following settings and it did not work
```
quarkusPlatformGroupId=io.quarkus
quarkusPluginVersion=1.3.1.Final
quarkusPlatformArtifactId=quarkus-bom
quarkusPlatformVersion=999-SNAPSHOT
```

@gastaldi please re-open this issue

```
gradle quarkusBuild -Dquarkus.container-image.build=true
[INFO] Initializing dekorate session.
[INFO] Default s2i build generator....
[INFO] Registering s2i handler!
[INFO] Registering s2i handler!
[INFO] Generating manifests.
[INFO] Processing kubernetes configuration.
[INFO] Processing openshift configuration.
[INFO] Processing s2i configuration.

> Task :quarkusBuild
building quarkus runner

[INFO] Closing dekorate session.

> Task :quarkusBuild FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':quarkusBuild'.
> io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.container.image.jib.deployment.JibProcessor#buildFromJar threw an exception: java.lang.NoSuchMethodError: com/google/common/base/Strings.isNullOrEmpty(Ljava/lang/String;)Z (loaded from jar:file:/Users/schintha/.m2/repository/org/jetbrains/kotlin/kotlin-compiler/1.3.71/kotlin-compiler-1.3.71.jar!/ by QuarkusClassLoader:Augmentation Class Loader) called from class com.google.cloud.tools.jib.api.ImageReference (loaded from jar:file:/Users/schintha/.m2/repository/com/google/cloud/tools/jib-core/0.13.1/jib-core-0.13.1.jar!/ by QuarkusClassLoader:Augmentation Class Loader).
        at com.google.cloud.tools.jib.api.ImageReference.parse(ImageReference.java:112)
        at io.quarkus.container.image.jib.deployment.JibProcessor.createContainerBuilderFromJar(JibProcessor.java:202)
        at io.quarkus.container.image.jib.deployment.JibProcessor.buildFromJar(JibProcessor.java:77)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:938)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:273)
        at io.quarkus.builder.BuildContext$$Lambda$163/0000000000000000.run(Unknown Source)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

--- a/extensions/kotlin/runtime/pom.xml
+++ b/extensions/kotlin/runtime/pom.xml
@@ -18,6 +18,14 @@
             <plugin>
                 <groupId>io.quarkus</groupId>
                 <artifactId>quarkus-bootstrap-maven-plugin</artifactId>
+                <configuration>
+                    <lesserPriorityArtifacts>
+                        <!--
+                        see https://github.com/quarkusio/quarkus/issues/8405
+                        -->
+                        <lesserPriorityArtifact>org.jetbrains.kotlin:kotlin-compiler</lesserPriorityArtifact>
+                    </lesserPriorityArtifacts>
+                </configuration>
             </plugin>
         </plugins>
     </build>

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapConstants.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/BootstrapConstants.java
@@ -21,6 +21,7 @@ public interface BootstrapConstants {
     String PROP_DEPLOYMENT_ARTIFACT = "deployment-artifact";
     String PARENT_FIRST_ARTIFACTS = "parent-first-artifacts";
     String EXCLUDED_ARTIFACTS = "excluded-artifacts";
+    String LESSER_PRIORITY_ARTIFACTS = "lesser-priority-artifacts";
 
     String EMPTY = "";
     String JAR = "jar";

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java
@@ -174,6 +174,8 @@ public class CuratedApplication implements Serializable, Closeable {
                     //we always load this from the parent if it is available, as this acts as a bridge between the running
                     //app and the dev mode code
                     builder.addParentFirstElement(element);
+                } else if (appModel.getLesserPriorityArtifacts().contains(key)) {
+                    builder.addLesserPriorityElement(element);
                 }
             }
             for (AppDependency userDep : appModel.getUserDependencies()) {
@@ -183,6 +185,8 @@ public class CuratedApplication implements Serializable, Closeable {
                     if (appModel.getParentFirstArtifacts().contains(key)) {
                         //this mostly happens when building quarkus itself
                         builder.addParentFirstElement(element);
+                    } else if (appModel.getLesserPriorityArtifacts().contains(key)) { // TODO: is this needed?
+                        builder.addLesserPriorityElement(element);
                     }
                     builder.addElement(element);
                 }
@@ -251,6 +255,8 @@ public class CuratedApplication implements Serializable, Closeable {
                     //we always load this from the parent if it is available, as this acts as a bridge between the running
                     //app and the dev mode code
                     builder.addParentFirstElement(element);
+                } else if (appModel.getLesserPriorityArtifacts().contains(key)) {
+                    builder.addLesserPriorityElement(element);
                 }
                 builder.addElement(element);
             }

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java
@@ -52,6 +52,7 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
     private final boolean aggregateParentResources;
     private final List<ClassPathElement> bannedElements;
     private final List<ClassPathElement> parentFirstElements;
+    private final List<ClassPathElement> lesserPriorityElements;
 
     /**
      * The element that holds resettable in-memory classses.
@@ -84,6 +85,7 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         this.bytecodeTransformers = builder.bytecodeTransformers;
         this.bannedElements = builder.bannedElements;
         this.parentFirstElements = builder.parentFirstElements;
+        this.lesserPriorityElements = builder.lesserPriorityElements;
         this.parent = builder.parent;
         this.parentFirst = builder.parentFirst;
         this.resettableElement = builder.resettableElement;
@@ -194,7 +196,24 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
                     }
                     Map<String, ClassPathElement[]> finalElements = new HashMap<>();
                     for (Map.Entry<String, List<ClassPathElement>> i : elementMap.entrySet()) {
-                        finalElements.put(i.getKey(), i.getValue().toArray(new ClassPathElement[i.getValue().size()]));
+                        List<ClassPathElement> entryClassPathElements = i.getValue();
+                        if (!lesserPriorityElements.isEmpty() && (entryClassPathElements.size() > 1)) {
+                            List<ClassPathElement> entryNormalPriorityElements = new ArrayList<>(entryClassPathElements.size());
+                            List<ClassPathElement> entryLesserPriorityElements = new ArrayList<>(entryClassPathElements.size());
+                            for (ClassPathElement classPathElement : entryClassPathElements) {
+                                if (lesserPriorityElements.contains(classPathElement)) {
+                                    entryLesserPriorityElements.add(classPathElement);
+                                } else {
+                                    entryNormalPriorityElements.add(classPathElement);
+                                }
+                            }
+                            // ensure the lesser priority elements are added later
+                            entryClassPathElements = new ArrayList<>(entryClassPathElements.size());
+                            entryClassPathElements.addAll(entryNormalPriorityElements);
+                            entryClassPathElements.addAll(entryLesserPriorityElements);
+                        }
+                        finalElements.put(i.getKey(),
+                                entryClassPathElements.toArray(new ClassPathElement[entryClassPathElements.size()]));
                     }
                     Set<String> banned = new HashSet<>();
                     for (ClassPathElement i : bannedElements) {
@@ -415,6 +434,7 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         final List<ClassPathElement> elements = new ArrayList<>();
         final List<ClassPathElement> bannedElements = new ArrayList<>();
         final List<ClassPathElement> parentFirstElements = new ArrayList<>();
+        final List<ClassPathElement> lesserPriorityElements = new ArrayList<>();
         Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers = Collections.emptyMap();
         final boolean parentFirst;
         MemoryClassPathElement resettableElement;
@@ -497,6 +517,19 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
          */
         public Builder addBannedElement(ClassPathElement element) {
             bannedElements.add(element);
+            return this;
+        }
+
+        /**
+         * Adds an element which will only be used to load a class or resource if no normal
+         * element containing that class or resource exists.
+         * This is used in order control the order of elements when multiple contain the same classes
+         *
+         * @param element The element to add
+         * @return This builder
+         */
+        public Builder addLesserPriorityElement(ClassPathElement element) {
+            lesserPriorityElements.add(element);
             return this;
         }
 

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/model/AppModel.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/model/AppModel.java
@@ -39,13 +39,17 @@ public class AppModel implements Serializable {
 
     private final Set<AppArtifactKey> parentFirstArtifacts;
 
+    private final Set<AppArtifactKey> lesserPriorityArtifacts;
+
     private AppModel(AppArtifact appArtifact, List<AppDependency> runtimeDeps, List<AppDependency> deploymentDeps,
-            List<AppDependency> fullDeploymentDeps, Set<AppArtifactKey> parentFirstArtifacts) {
+            List<AppDependency> fullDeploymentDeps, Set<AppArtifactKey> parentFirstArtifacts,
+            Set<AppArtifactKey> lesserPriorityArtifacts) {
         this.appArtifact = appArtifact;
         this.runtimeDeps = runtimeDeps;
         this.deploymentDeps = deploymentDeps;
         this.fullDeploymentDeps = fullDeploymentDeps;
         this.parentFirstArtifacts = parentFirstArtifacts;
+        this.lesserPriorityArtifacts = lesserPriorityArtifacts;
     }
 
     public AppArtifact getAppArtifact() {
@@ -76,6 +80,10 @@ public class AppModel implements Serializable {
         return parentFirstArtifacts;
     }
 
+    public Set<AppArtifactKey> getLesserPriorityArtifacts() {
+        return lesserPriorityArtifacts;
+    }
+
     @Override
     public String toString() {
         return "AppModel{" +
@@ -96,6 +104,7 @@ public class AppModel implements Serializable {
         private final List<AppDependency> runtimeDeps = new ArrayList<>();
         private final Set<AppArtifactKey> parentFirstArtifacts = new HashSet<>();
         private final Set<AppArtifactKey> excludedArtifacts = new HashSet<>();
+        private final Set<AppArtifactKey> lesserPriorityArtifacts = new HashSet<>();
 
         public Builder setAppArtifact(AppArtifact appArtifact) {
             this.appArtifact = appArtifact;
@@ -173,6 +182,14 @@ public class AppModel implements Serializable {
                     log.debugf("Extension %s is excluding %s", extension, artifact);
                 }
             }
+            String lesserPriority = props.getProperty(BootstrapConstants.LESSER_PRIORITY_ARTIFACTS);
+            if (lesserPriority != null) {
+                String[] artifacts = lesserPriority.split(",");
+                for (String artifact : artifacts) {
+                    lesserPriorityArtifacts.add(new AppArtifactKey(artifact.split(":")));
+                    log.debugf("Extension %s is making %s a lesser priority artifact", extension, artifact);
+                }
+            }
         }
 
         public AppModel build() {
@@ -192,7 +209,7 @@ public class AppModel implements Serializable {
             List<AppDependency> fullDeploymentDeps = this.fullDeploymentDeps.stream().filter(includePredicate)
                     .collect(Collectors.toList());
             AppModel appModel = new AppModel(appArtifact, runtimeDeps, deploymentDeps, fullDeploymentDeps,
-                    parentFirstArtifacts);
+                    parentFirstArtifacts, lesserPriorityArtifacts);
             log.debugf("Created AppMode %s", appModel);
             return appModel;
 

--- a/independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java
+++ b/independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java
@@ -108,6 +108,14 @@ public class ExtensionDescriptorMojo extends AbstractMojo {
     @Parameter
     List<String> parentFirstArtifacts;
 
+    /**
+     * Artifacts that will only be used to load a class or resource if no other normal element exists.
+     * This is an advanced option that should only be used when there is a case of multiple jars
+     * containing the same classes and we need to control which jars is actually used to load the classes.
+     */
+    @Parameter
+    List<String> lesserPriorityArtifacts;
+
     @Override
     public void execute() throws MojoExecutionException {
 
@@ -126,6 +134,11 @@ public class ExtensionDescriptorMojo extends AbstractMojo {
         if (excludedArtifacts != null && !excludedArtifacts.isEmpty()) {
             String val = String.join(",", excludedArtifacts);
             props.put(BootstrapConstants.EXCLUDED_ARTIFACTS, val);
+        }
+
+        if (lesserPriorityArtifacts != null && !lesserPriorityArtifacts.isEmpty()) {
+            String val = String.join(",", lesserPriorityArtifacts);
+            props.put(BootstrapConstants.LESSER_PRIORITY_ARTIFACTS, val);
         }
 
         try {

